src/js/components/conversation.jsx

```diff
@@ -54,34 +65,145 @@ export class ConversationComponent extends Component {
 
     };
 
+    onScroll = () => {
+        const {dispatch, shouldScrollToBottom} = this.props;
+
+        // If top of Conversation component is reached, we need to fetch older messages
+        const node = findDOMNode(this);
+        if(node.scrollTop === 0) {
+            this.fetchHistory();
+        } else if(shouldScrollToBottom) {
+            dispatch(setShouldScrollToBottom(false));
+        }
+    };
+
+    fetchHistory = () => {
+        const {dispatch, hasMoreMessages, isFetchingMoreMessages, messages} = this.props;
+
+        const node = findDOMNode(this);
+        if(hasMoreMessages && !isFetchingMoreMessages) {
+            // make sure the last message is one from the server, otherwise it doesn't need to scroll to previous first message
+            if (messages.length > 0 && messages[messages.length - 1]._id) {
+                this._lastTopMessageId = messages[0]._id;
+            }
+            
+            const top = getTop(this._topMessageNode, node);
+            this._lastTopMessageNodePosition = top - node.scrollTop;
+            dispatch(setFetchingMoreMessages(true));
+            setTimeout(() => {
```

Comment:
You usually use setTimeout when you need to compute sizes because the
function in setTimeout will be executed when the html is effectively
flushed to the dom so it would reflect the real sizes with css and
everything

On Aug 18, 2016 3:54 PM, "chloepouprom" <notifications@github.com> wrote:

> In src/js/components/conversation.jsx
> <https://github.com/smooch/smooch-js/pull/374#discussion_r75377305>:
>
> > +    };
> > +
> > +    fetchHistory = () => {
> > +        const {dispatch, hasMoreMessages, isFetchingMoreMessages, messages} = this.props;
> > +
> > +        const node = findDOMNode(this);
> > +        if(hasMoreMessages && !isFetchingMoreMessages) {
> > +            // make sure the last message is one from the server, otherwise it doesn't need to scroll to previous first message
> > +            if (messages.length > 0 && messages[messages.length - 1]._id) {
> > +                this._lastTopMessageId = messages[0]._id;
> > +            }
> > +
> > +            const top = getTop(this._topMessageNode, node);
> > +            this._lastTopMessageNodePosition = top - node.scrollTop;
> > +            dispatch(setFetchingMoreMessages(true));
> > +            setTimeout(() => {
>
> I think this was originally so that the transition from the Load more to Retrieving
> history text appears less choppy. But now, looks like it doesn't actually
> make a difference... I'll remove this
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/smooch/smooch-js/pull/374/files/81170bcbd64e9da1408e20b8b55912cd25985932#r75377305>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAvuFOQgU0bn54H55UH2_Hjf8e3-q1FGks5qhLh8gaJpZM4Jm5kw>
> .
>